saleor/checkout/steps.py

```diff
@@ -23,198 +23,161 @@ def __init__(self, request, storage):
-        address_data = storage.get('address', {})
-        address = Address(**address_data)
-        skip = False
-        if not address_data and request.user.is_authenticated():
-            if request.user.default_billing_address:
-                address = request.user.default_billing_address
-                skip = True
-            elif request.user.addresses.count() == 1:
-                address = request.user.addresses.all()[0].address
-                skip = True
-        super(BillingAddressStep, self).__init__(request, storage, address)
-        if not request.user.is_authenticated():
-            self.anonymous_user_email = self.storage.get(
-                'anonymous_user_email')
-            initial = {'email': self.anonymous_user_email}
-            self.forms['anonymous'] = AnonymousEmailForm(request.POST or None,
-                                                         initial=initial)
-        else:
-            self.anonymous_user_email = ''
-        if skip:
-            self.save()
-
-    def __str__(self):
-        return 'billing-address'
-
-    def forms_are_valid(self):
-        forms_are_valid = super(BillingAddressStep, self).forms_are_valid()
-        if 'anonymous' not in self.forms:
-            return forms_are_valid
-        anonymous_form = self.forms['anonymous']
-        if forms_are_valid and anonymous_form.is_valid():
-            self.anonymous_user_email = anonymous_form.cleaned_data['email']
-            return True
-        return False
+        return super(BaseCheckoutStep, self).process(extra_context=context)
 
     def save(self):
-        self.storage['anonymous_user_email'] = self.anonymous_user_email
+        if not self.authenticated_user:
+            self.email = self.forms['email'].cleaned_data['email']
+        self.storage['email'] = self.email
+        address = self.forms['address'].cleaned_data
+        self.address = Address(**address)
         self.storage['address'] = Address.objects.as_data(self.address)
 
     def add_to_order(self, order):
         self.address.save()
-        order.anonymous_user_email = self.anonymous_user_email
-        order.billing_address = self.address
+        order.shipping_address = self.address
         if order.user:
-            User.objects.store_address(order.user, self.address, billing=True)
+            User.objects.store_address(order.user, self.address, shipping=True)
 
     def validate(self):
-        super(BillingAddressStep, self).validate()
-        if 'anonymous' in self.forms and not self.anonymous_user_email:
+        try:
+            self.address.clean_fields()
+        except ValidationError as e:
+            raise InvalidData(e.messages)
+
+        if not self.email:
             raise InvalidData()
 
+    def forms_are_valid(self):
+        if self.forms['address'].is_valid() and (
+            self.authenticated_user or self.forms['email'].is_valid()):
+            return True
+        return False
```

Comment:
You don't have to construct whole `if/else` statement here, just return condition result